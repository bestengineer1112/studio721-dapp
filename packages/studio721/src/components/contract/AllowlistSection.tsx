import { Address } from '@openpalette/contract';
import { PlusIcon } from '@radix-ui/react-icons';
import {
  HStack,
  Small,
  SpacerHorizontal,
  FormSection,
  InfoHoverCard,
} from 'components';
import { ArrayController, Button, IconButton, InputField } from 'designsystem';
import React, { memo, ReactNode, useCallback } from 'react';
import { AllowlistDestination, Action } from 'state';

export const DestinationRow = memo(function DestinationRow({
  id,
  address,
  amount,
  disabled,
  addressDisabled,
  onClickCross,
  onChangeAddress,
  onChangeAmount,
}: {
  id: string;
  address: Address;
  amount: number;
  disabled: boolean;
  addressDisabled?: boolean;
  onClickCross: () => void;
  onChangeAddress: (value: Address) => void;
  onChangeAmount: (value: number) => void;
}) {
  return (
    <HStack alignItems="center" flex="1 1 0px">
      <IconButton
        iconName="DragHandleDots2Icon"
        color={disabled || addressDisabled ? 'transparent' : undefined}
      />
      <SpacerHorizontal size={6} />
      <InputField.Root id={`${id}-amount`} flex="1 1 0px" labelSize={12}>
        <InputField.NumberInput
          autoComplete="no"
          disabled={disabled}
          value={amount}
          onSubmit={onChangeAmount}
          onChangeInternalValue={useCallback(
            (value) => {
              const numberValue = Number(value);

              if (
                !Number.isInteger(numberValue) ||
                numberValue < 0 ||
                numberValue > 100
              )
                return;

              onChangeAmount(numberValue);
            },
            [onChangeAmount],
          )}
          onNudge={useCallback(
            (delta) => {
              onChangeAmount(Math.max(0, Math.min(100, amount + delta)));
            },
            [amount, onChangeAmount],
          )}
        />
      </InputField.Root>
      <SpacerHorizontal size={10} />
      <InputField.Root id={`${id}-address`} flex="3 3 0px">
        <InputField.Input
          disabled={disabled || addressDisabled}
          value={address}
          onChange={onChangeAddress as (value: string) => void}
        />
      </InputField.Root>
      <SpacerHorizontal size={6} />
      <IconButton
        iconName="Cross2Icon"
        color={addressDisabled || disabled ? 'transparent' : undefined}
        onClick={onClickCross}
      />
    </HStack>
  );
});

interface Props {
  destinations: AllowlistDestination[];
  allowedForOwner: number;
  disabled: boolean;
  dispatch: (action: Action) => void;
}

export const AllowlistSection = memo(function AllowlistSection({
  allowedForOwner,
  destinations,
  disabled,
  dispatch,
}: Props) {
  const handleMoveItem = useCallback(
    (sourceIndex: number, destinationIndex: number) => {
      dispatch({
        type: 'moveAllowlistDestination',
        sourceIndex,
        destinationIndex,
      });
    },
    [dispatch],
  );

  const handleRenderItem = useCallback(
    ({
      item,
      index,
    }: {
      item: AllowlistDestination;
      index: number;
    }): ReactNode => {
      return (
        <DestinationRow
          id={`allowlist-destination-${index}`}
          key={index}
          address={item.address}
          amount={item.amount}
          disabled={disabled}
          onChangeAmount={(amount) => {
            if (amount === item.amount) return;

            dispatch({ type: 'setAllowlistDestinationAmount', index, amount });
          }}
          onChangeAddress={(address) => {
            if (address === item.address) return;

            dispatch({
              type: 'setAllowlistDestinationAddress',
              index,
              address,
            });
          }}
          onClickCross={() => {
            dispatch({ type: 'removeAllowlistDestination', index });
          }}
        />
      );
    },
    [disabled, dispatch],
  );

  return (
    <FormSection
      title={
        <>
          Allowlist{' '}
          <InfoHoverCard top="1px">
            Addresses allowed to mint, and how many they can mint, before the
            sale starts.
          </InfoHoverCard>
        </>
      }
      right={
        <Button
          disabled={disabled}
          onClick={useCallback(() => {
            dispatch({ type: 'addAllowlistDestination' });
          }, [dispatch])}
        >
          Add Address
          <SpacerHorizontal inline size={4} />
          <PlusIcon />
        </Button>
      }
    >
      <HStack gap={3}>
        <HStack flex="1 1 0px" minWidth={0} padding={'0 0 0 21px'}>
          <Small>Amount</Small>
        </HStack>
        <HStack flex="3 3 0px" minWidth={0} padding={'0 0 0 0'}>
          <Small>Allowed Address</Small>
        </HStack>
      </HStack>
      <DestinationRow
        id={`allowlist-destination-owner`}
        address={'Contract Owner' as any}
        amount={allowedForOwner}
        disabled={disabled}
        addressDisabled={true}
        onChangeAmount={(amount) => {
          if (amount === allowedForOwner) return;

          dispatch({ type: 'setAmountAllowedForOwner', amount });
        }}
        onChangeAddress={(address) => {}}
        onClickCross={() => {}}
      />
      <ArrayController<AllowlistDestination>
        id={'allowlist-destination-controller'}
        sortable
        items={destinations}
        onMoveItem={handleMoveItem}
        renderItem={handleRenderItem}
      />
    </FormSection>
  );
});
