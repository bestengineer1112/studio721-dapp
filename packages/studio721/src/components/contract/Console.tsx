import { Divider, SpacerVertical, VStack } from 'components';
import React, { ReactNode, useEffect, useRef } from 'react';
import { withSeparatorElements } from 'utils';

export function Console({ elements }: { elements: ReactNode[] }) {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    ref.current?.scrollTo({
      top: 99999,
    });
  }, [elements]);

  return (
    <VStack
      ref={ref}
      overflowY="auto"
      flex="1 1 0px"
      padding={20}
      background="rgba(255,255,255,0.06)"
    >
      {withSeparatorElements(
        elements,
        <>
          <SpacerVertical size={20} />
          <Divider variant="light" />
          <SpacerVertical size={20} />
        </>,
      )}
    </VStack>
  );
}
