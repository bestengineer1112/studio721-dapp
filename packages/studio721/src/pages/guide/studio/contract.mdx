import { SpacerHorizontal } from 'components';

Studio 721 Contract is a free tool for configuring, compiling, deploying, and
verifying custom [ERC-721](https://eips.ethereum.org/EIPS/eip-721) NFT smart
contracts.

## Features

- 🤯<SpacerHorizontal inline size={20} />Create a custom NFT contract without
  writing code
- 🚀<SpacerHorizontal inline size={20} />Compile, deploy, and verify all without
  leaving your browser!
- ⚙️<SpacerHorizontal inline size={20} />Lots of configuration options
- 🏠<SpacerHorizontal inline size={20} />Your config is stored in your browser
  url (no centralized servers)
- 💸<SpacerHorizontal inline size={20} />Gas-optimized deployment to save you
  money
- 💻<SpacerHorizontal inline size={20} />Royalties (ERC-2981) and multiple
  payout addresses
- 👩‍💻<SpacerHorizontal inline size={20} />You can still download and customize
  the code if you want
- 🔓<SpacerHorizontal inline size={20} />Free and open source!

## Walkthrough Video

This is part 3 of a
[4-part video series](https://youtu.be/iWV5c_HWUw8?list=PLu07TXQzA0jo9RuBstWDDmm1LOorye3aP):

<YouTube src="https://www.youtube.com/embed/4p89heWL5VM?list=PLu07TXQzA0jo9RuBstWDDmm1LOorye3aP" />
