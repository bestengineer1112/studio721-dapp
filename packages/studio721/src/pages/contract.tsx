import { Interface } from '@ethersproject/abi';
import { Address, CHAIN_ID, getChainName } from '@openpalette/contract';
import {
  Button,
  CodeHighlight,
  ConnectButton,
  Divider,
  getHeadTags,
  HStack,
  LinkChip,
  Regular,
  SpacerHorizontal,
  SpacerVertical,
  TwitterChip,
  VStack,
} from 'components';
import { useChainId, useWeb3Data } from 'contexts';
import { usePrismExtensions } from 'hooks';
import Head from 'next/head';
import { useRouter } from 'next/router';
import theme from 'prism-react-renderer/themes/vsDark';
import React, {
  useCallback,
  useEffect,
  useMemo,
  useReducer,
  useRef,
} from 'react';
import {
  downloadDependenciesForSource,
  generateContractSource,
  getBaseURI,
  getValidContractName,
} from 'solidity-codegen';
import { createCompilerInput } from 'solidity-compiler';
import {
  ContractConfigState,
  createInitialState,
  GWEI_SLIM_VERSION,
  OPEN_ZEPPELIN_VERSION,
  reducer,
  ScopedAccessToken,
} from 'state';
import { SolidityCompilerOutput, WorkerRequest, WorkerResponse } from 'utils';
import { getEtherscanApiUrl } from 'web3-utils';
import { AllowlistSection } from '../components/contract/AllowlistSection';
import { ContractConsole } from '../components/contract/ContractConsole';
import { MintingSection } from '../components/contract/MintingSection';
import { ParametersSection } from '../components/contract/ParametersSection';
import { PayoutSection } from '../components/contract/PayoutSection';
import { TokenPreview } from '../components/contract/TokenPreview';
import { VerificationSection } from '../components/contract/VerificationSection';
import { deployContract } from '../utils/deploy';
import { socialConfig } from '../utils/socialConfig';
import {
  checkVerificationStatus,
  toCheckStatusRequest,
  toVerifyRequest,
  verifyContract,
} from '../utils/verify';

const SCHEMA_VERSION_KEY = 'schemaVersion';

try {
  theme.plain.backgroundColor = 'black';
} catch {
  //
}

const erc721DelegatedAddress: ScopedAccessToken = {
  mainnet: '0x43955024b1985e2b933a59021500ae5f55b04091',
  rinkeby: '0x86c67a16c16bf784bdfe7d4b7575db664d191f88',
  ropsten: '',
  goerli: '',
  polygon: '',
  mumbai: '',
};

function useCompiler() {
  const requestId = useRef(0);
  const worker = useRef<Worker>();

  useEffect(() => {
    worker.current = new Worker(new URL('../worker.ts', import.meta.url));
  }, [worker]);

  return {
    compile: (files: Record<string, string>) => {
      return new Promise<SolidityCompilerOutput>((resolve) => {
        const request: WorkerRequest = {
          id: requestId.current++,
          type: 'compile',
          request: {
            input: createCompilerInput(files),
          },
        };

        function handler(e: MessageEvent) {
          const data = e.data as WorkerResponse;

          if (!(data.id === request.id && data.type === request.type)) return;

          worker.current?.removeEventListener('message', handler);

          resolve(data.response.output);
        }

        worker.current?.addEventListener('message', handler);
        worker.current?.postMessage(request);
      });
    },
  };
}

export default function NFTStudio() {
  usePrismExtensions();

  const router = useRouter();

  const initialConfig = useMemo(() => {
    if (typeof router.query.config !== 'string') return {};

    const config = router.query.config
      ? JSON.parse(decodeURIComponent(router.query.config))
      : {};

    delete config[SCHEMA_VERSION_KEY];

    return config;
  }, [router.query.config]);

  const [state, dispatch] = useReducer(
    reducer,
    createInitialState(initialConfig),
  );

  const configJSON = useMemo(() => {
    const clone = { ...state.config };

    (
      Object.entries(createInitialState({}).config) as [
        keyof ContractConfigState,
        any,
      ][]
    ).forEach(([key, value]) => {
      if (JSON.stringify(clone[key]) === JSON.stringify(value)) {
        delete clone[key];
      }
    });

    return JSON.stringify({
      ...clone,
      [SCHEMA_VERSION_KEY]: '1.0.0',
    });
  }, [state.config]);

  useEffect(() => {
    const config = `config=${encodeURIComponent(configJSON)}`;

    if (router.asPath.includes(config)) return;

    router.replace(
      {
        query: config,
      },
      undefined,
      {
        scroll: false,
        shallow: true,
      },
    );
  }, [configJSON, router]);

  const compiler = useCompiler();
  const provider = useWeb3Data()?.provider;
  const chainId = useChainId();

  const handleCompile = useCallback(() => {
    async function main() {
      const contractName = getValidContractName(state.config.tokenName);
      const sourceName = contractName + '.sol';

      const source = generateContractSource(state.config);

      dispatch({ type: 'downloadFiles' });

      const files = await downloadDependenciesForSource(
        fetch,
        sourceName,
        source,
        {
          '@openzeppelin/contracts': OPEN_ZEPPELIN_VERSION,
          'gwei-slim-nft-contracts': GWEI_SLIM_VERSION,
        },
      );

      dispatch({ type: 'setCompilerReady', value: files });

      const output = await compiler.compile(files);

      dispatch({
        type: 'setCompilerDone',
        value: output,
        sourceName,
        contractName,
      });
    }

    main();
  }, [compiler, state.config]);

  const handleDeploy = useCallback(() => {
    async function main() {
      if (state.compiler.type !== 'done' || !provider || !chainId) return;

      const { sourceName, contractName, contracts } = state.compiler;

      const mainContract = contracts[sourceName][contractName];

      const {
        abi,
        evm: { bytecode },
      } = mainContract;

      dispatch({
        type: 'setDeploying',
      });

      let deploymentAddress: Address;

      const chainName = getChainName(chainId);

      try {
        if (
          state.config.usesDelegatedContract &&
          !(chainId === CHAIN_ID.MAINNET || chainId === CHAIN_ID.RINKEBY)
        ) {
          alert(
            "You can't use the 'Reduce deployment costs' option on this network! Mainnet and Rinkeby only for now.",
          );

          throw new Error("Can't use delegated contracts on this network");
        }

        deploymentAddress = await deployContract(
          provider.getSigner(),
          abi,
          bytecode,
          [
            ...(state.config.usesDelegatedContract
              ? [erc721DelegatedAddress[chainName] ?? undefined]
              : []),
            getBaseURI(state.config.tokenURI),
            ...(state.config.requireAccessToken
              ? [state.config.requireAccessToken[chainName]]
              : []),
            ...(state.config.approvalProxyAddress
              ? [state.config.approvalProxyAddress[chainName]]
              : []),
          ],
        );
      } catch (e) {
        console.warn('deploy failure', e);

        dispatch({
          type: 'deployFailure',
        });

        return;
      }

      dispatch({
        type: 'setDeploymentAddress',
        value: deploymentAddress,
      });
    }

    main();
  }, [
    state.compiler,
    state.config.usesDelegatedContract,
    state.config.tokenURI,
    state.config.requireAccessToken,
    state.config.approvalProxyAddress,
    provider,
    chainId,
  ]);

  const handleVerify = useCallback(() => {
    async function main() {
      if (
        state.compiler.type !== 'done' ||
        state.deployment.type !== 'deployed' ||
        !chainId
      )
        return;

      const { contracts, contractName, sourceName } = state.compiler;

      const contractInterface = new Interface(
        contracts[sourceName][contractName].abi,
      );

      const chainName = getChainName(chainId);

      const deployArguments = contractInterface
        .encodeDeploy([
          ...(state.config.usesDelegatedContract
            ? [erc721DelegatedAddress[chainName] ?? undefined]
            : []),
          getBaseURI(state.config.tokenURI),
          ...(state.config.requireAccessToken
            ? [state.config.requireAccessToken[chainName]]
            : []),
          ...(state.config.approvalProxyAddress
            ? [state.config.approvalProxyAddress[chainName]]
            : []),
        ])
        .replace('0x', '');

      const verifyRequest = toVerifyRequest({
        apiKey: state.apiKeys.etherscan,
        compilerVersion: 'v0.8.9+commit.e5eed63a',
        constructorArguments: deployArguments,
        contractAddress: state.deployment.address,
        contractName: state.compiler.contractName,
        sourceCode: JSON.stringify(createCompilerInput(state.compiler.files)),
        sourceName: state.compiler.contractName + '.sol',
      });

      dispatch({ type: 'setVerifying' });

      // console.log('req', verifyRequest);

      const verificationResult = await verifyContract(
        getEtherscanApiUrl(chainId),
        verifyRequest,
      );

      if (verificationResult instanceof Error) {
        dispatch({ type: 'setVerificationFailed', error: verificationResult });

        return;
      }

      // console.log('res', verificationResult);

      const checkStatusRequest = toCheckStatusRequest({
        apiKey: state.apiKeys.etherscan,
        guid: verificationResult.result,
      });

      const checkStatusResult = await checkVerificationStatus(
        getEtherscanApiUrl(chainId),
        checkStatusRequest,
      );

      if (checkStatusResult instanceof Error) {
        console.warn('Failed to verify', checkStatusResult);

        dispatch({ type: 'setVerificationFailed', error: checkStatusResult });
      } else {
        dispatch({ type: 'setVerified' });
      }
    }

    main();
  }, [
    chainId,
    state.apiKeys.etherscan,
    state.compiler,
    state.config.approvalProxyAddress,
    state.config.tokenURI,
    state.config.requireAccessToken,
    state.config.usesDelegatedContract,
    state.deployment,
  ]);

  const hasStartedDeploying = state.deployment.type !== 'notStarted';

  useEffect(() => {
    localStorage.setItem('etherscanApiKey', state.apiKeys.etherscan);
  }, [state.apiKeys.etherscan]);

  return (
    <>
      <Head>
        <title>Studio 721 Contract</title>
        {getHeadTags({
          pageTitle: 'Studio 721 Contract',
          pageDescription:
            'The free, all-in-one tool for configuring, deploying, and verifying NFT smart contracts.',
          config: socialConfig,
        })}
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <HStack
        margin={'60px 0 0 0'}
        flex={'1 1 0%'}
        breakpoints={{
          [800]: {
            flexDirection: 'column',
          },
        }}
      >
        <VStack
          flex="1 1 0px"
          breakpoints={{
            [800]: {
              flex: '0 0 auto',
            },
          }}
        >
          <VStack
            overflowY="auto"
            flex="1 1 0px"
            gap={40}
            padding={'20px 40px'}
            breakpoints={{
              [800]: {
                overflowY: 'initial',
              },
            }}
          >
            <VStack gap={10}>
              <Regular>
                <strong>Studio 721</strong> is a free tool for configuring,
                compiling, deploying, and verifying custom{' '}
                <LinkChip
                  openInNewTab
                  href="https://docs.openzeppelin.com/contracts/2.x/api/token/erc721"
                  style={{
                    whiteSpace: 'pre',
                  }}
                >
                  ERC 721
                </LinkChip>{' '}
                NFT smart contracts. Studio 721 doesn{"'"}t host your assets or
                metadata; use your favorite hosting service and link it to your
                NFT via {'"'}
                Token URI{'"'}. Not sure where to start? There's a{' '}
                <LinkChip href="/guide/studio/contract">
                  video walkthrough!
                </LinkChip>
              </Regular>
              <Regular>
                Created by <TwitterChip value="@dvnabbott" />
              </Regular>
            </VStack>
            <TokenPreview
              config={state.config}
              disabled={hasStartedDeploying}
              dispatch={dispatch}
            />
            <MintingSection
              config={state.config}
              disabled={hasStartedDeploying}
              dispatch={dispatch}
            />
            <ParametersSection
              disabled={hasStartedDeploying}
              dispatch={dispatch}
              parameters={state.config.tokenParameters}
            />
            <PayoutSection
              disabled={hasStartedDeploying}
              dispatch={dispatch}
              destinations={state.config.payoutDestinations}
            />
            <AllowlistSection
              allowedForOwner={state.config.amountAllowedForOwner}
              disabled={hasStartedDeploying}
              dispatch={dispatch}
              destinations={state.config.allowlistDestinations}
            />
            <VerificationSection
              etherscanApiKey={state.apiKeys.etherscan}
              dispatch={dispatch}
            />
            <SpacerVertical size={10} />
          </VStack>
        </VStack>
        <VStack
          flex="1 1 0%"
          background="black"
          breakpoints={{
            [800]: {
              flex: '0 0 auto',
            },
          }}
        >
          <VStack
            overflowY="auto"
            flex="2 2 0px"
            breakpoints={{
              [800]: {
                minHeight: 400,
              },
            }}
          >
            <VStack flex="1 1 0px" padding={20}>
              <CodeHighlight
                code={generateContractSource(state.config)}
                language={'solidity' as any}
              />
            </VStack>
          </VStack>
          <Divider variant="light" />
          <VStack
            overflowY="auto"
            flex="1 1 0px"
            breakpoints={{
              [800]: {
                minHeight: 400,
              },
            }}
          >
            <ContractConsole
              compiler={state.compiler}
              deployment={state.deployment}
              verification={state.verification}
            />
          </VStack>
          <Divider variant="light" />
          <VStack gap={20} padding={20}>
            <HStack gap={20} flexWrap="wrap">
              <Button
                disabled={state.compiler.type === 'done'}
                onClick={handleCompile}
              >
                Compile
              </Button>
              <Button
                disabled={
                  !provider ||
                  state.compiler.type !== 'done' ||
                  state.deployment.type !== 'notStarted'
                }
                onClick={handleDeploy}
              >
                {chainId
                  ? `Deploy on ${getChainName(chainId)}`
                  : 'Deploy' +
                    (state.compiler.type === 'done' && !provider
                      ? ' (connect wallet to deploy)'
                      : '')}
              </Button>
              <Button
                disabled={
                  state.deployment.type !== 'deployed' ||
                  state.verification.type !== 'notStarted' ||
                  !state.apiKeys.etherscan
                }
                onClick={handleVerify}
              >
                Verify{' '}
                {!state.apiKeys.etherscan &&
                state.deployment.type === 'deployed'
                  ? '(need API key)'
                  : ''}
              </Button>
              <SpacerHorizontal />
              <HStack alignItems="center">
                <ConnectButton />
              </HStack>
            </HStack>
          </VStack>
        </VStack>
      </HStack>
    </>
  );
}

// Next router only picks up parameters if we render server-side
export async function getServerSideProps(context: any) {
  return { props: {} };
}
