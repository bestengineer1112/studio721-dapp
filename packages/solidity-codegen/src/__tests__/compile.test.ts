import fetch from 'cross-fetch';
import {
  ContractConfigState,
  createInitialState,
  GWEI_SLIM_VERSION,
  OPEN_ZEPPELIN_VERSION,
} from 'state';
import { downloadDependenciesForSource, generateContractSource } from '..';
import {
  createCompilerInput,
  getCompiler,
  SolidityCompiler,
} from 'solidity-compiler';
import { createAddress } from '@openpalette/contract';

jest.setTimeout(30000);

let compiler: SolidityCompiler;

beforeAll(async () => {
  compiler = await getCompiler();
});

const libraryVersions = {
  '@openzeppelin/contracts': OPEN_ZEPPELIN_VERSION,
  'gwei-slim-nft-contracts': GWEI_SLIM_VERSION,
};

const complexInput: Partial<ContractConfigState> = {
  tokenURI:
    'https://www.721.so/api/example/metadata/{tokenId}.json{parameters}',
  multimint: 20,
  price: '0.02',
  activateAutomatically: true,
  usesUriStorage: true,
  usesIdParameter: true,
  requireAccessToken: {
    mainnet: '0x1308c158e60D7C4565e369Df2A86eBD853EeF2FB',
    rinkeby: '0x6C989C4Eda8E3fABce543Af5bfaa0D67b256354e',
    ropsten: '',
    goerli: '',
    polygon: '',
    mumbai: '',
  },
  tokenParameters: [{ name: 'param0', type: 'uint256' }],
  payoutDestinations: [
    {
      address: createAddress('0x0000000000000000000000000000000000000000'),
      amount: 50,
    },
  ],
  allowlistDestinations: [
    {
      address: createAddress('0x0000000000000000000000000000000000000000'),
      amount: 1,
    },
  ],
  amountAllowedForOwner: 2,
  approvalProxyAddress: {
    mainnet: '0xa5409ec958c83c3f309868babaca7c86dcb077c1',
    rinkeby: '0xf57b2c51ded3a29e6891aba85459d600256cf317',
    ropsten: '',
    goerli: '',
    polygon: '',
    mumbai: '',
  },
  limitPerWallet: 5,
  royaltyBps: '10',
};

it('compiles', async () => {
  const state = createInitialState({});

  const source = generateContractSource(state.config);

  const files = await downloadDependenciesForSource(
    fetch,
    'Test.sol',
    source,
    libraryVersions,
  );

  const compiled = compiler.compile(createCompilerInput(files));

  expect(Object.keys(compiled.sources)).toMatchSnapshot();
});

it('compiles complex', async () => {
  const state = createInitialState(complexInput);

  const source = generateContractSource(state.config);

  const files = await downloadDependenciesForSource(
    fetch,
    'Test.sol',
    source,
    libraryVersions,
  );

  const compiled = compiler.compile(createCompilerInput(files));

  expect(Object.keys(compiled.sources)).toMatchSnapshot();
});

it('compiles + delegation', async () => {
  const state = createInitialState({
    usesDelegatedContract: true,
  });

  const source = generateContractSource(state.config);

  const files = await downloadDependenciesForSource(
    fetch,
    'Test.sol',
    source,
    libraryVersions,
  );

  const compiled = compiler.compile(createCompilerInput(files));

  expect(Object.keys(compiled.sources)).toMatchSnapshot();
});

it('compiles complex + delegation', async () => {
  const state = createInitialState({
    ...complexInput,
    usesDelegatedContract: true,
  });

  const source = generateContractSource(state.config);

  const files = await downloadDependenciesForSource(
    fetch,
    'Test.sol',
    source,
    libraryVersions,
  );

  const compiled = compiler.compile(createCompilerInput(files));

  expect(Object.keys(compiled.sources)).toMatchSnapshot();
});
