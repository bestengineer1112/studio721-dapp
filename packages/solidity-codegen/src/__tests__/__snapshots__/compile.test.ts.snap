// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles + delegation 1`] = `
Array [
  "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/interfaces/IERC165Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/interfaces/IERC2981Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721ReceiverUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/IERC721MetadataUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/StorageSlotUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/StringsUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/introspection/IERC165Upgradeable.sol",
  "@openzeppelin/contracts/security/ReentrancyGuard.sol",
  "@openzeppelin/contracts/utils/Counters.sol",
  "Test.sol",
  "gwei-slim-nft-contracts/contracts/base/ERC721Base.sol",
  "gwei-slim-nft-contracts/contracts/base/ERC721Delegated.sol",
  "gwei-slim-nft-contracts/contracts/base/IBaseERC721Interface.sol",
]
`;

exports[`compiles 1`] = `
Array [
  "@openzeppelin/contracts/access/Ownable.sol",
  "@openzeppelin/contracts/security/ReentrancyGuard.sol",
  "@openzeppelin/contracts/token/ERC721/ERC721.sol",
  "@openzeppelin/contracts/token/ERC721/IERC721.sol",
  "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol",
  "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol",
  "@openzeppelin/contracts/utils/Address.sol",
  "@openzeppelin/contracts/utils/Context.sol",
  "@openzeppelin/contracts/utils/Counters.sol",
  "@openzeppelin/contracts/utils/Strings.sol",
  "@openzeppelin/contracts/utils/introspection/ERC165.sol",
  "@openzeppelin/contracts/utils/introspection/IERC165.sol",
  "Test.sol",
]
`;

exports[`compiles complex + delegation 1`] = `
Array [
  "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/interfaces/IERC165Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/interfaces/IERC2981Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721ReceiverUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/IERC721MetadataUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/StorageSlotUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/StringsUpgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol",
  "@openzeppelin/contracts-upgradeable/utils/introspection/IERC165Upgradeable.sol",
  "@openzeppelin/contracts/security/ReentrancyGuard.sol",
  "@openzeppelin/contracts/token/ERC721/IERC721.sol",
  "@openzeppelin/contracts/utils/Address.sol",
  "@openzeppelin/contracts/utils/Counters.sol",
  "@openzeppelin/contracts/utils/Strings.sol",
  "@openzeppelin/contracts/utils/introspection/IERC165.sol",
  "Test.sol",
  "gwei-slim-nft-contracts/contracts/base/ERC721Base.sol",
  "gwei-slim-nft-contracts/contracts/base/ERC721Delegated.sol",
  "gwei-slim-nft-contracts/contracts/base/IBaseERC721Interface.sol",
]
`;

exports[`compiles complex 1`] = `
Array [
  "@openzeppelin/contracts/access/Ownable.sol",
  "@openzeppelin/contracts/interfaces/IERC165.sol",
  "@openzeppelin/contracts/interfaces/IERC2981.sol",
  "@openzeppelin/contracts/security/ReentrancyGuard.sol",
  "@openzeppelin/contracts/token/ERC721/ERC721.sol",
  "@openzeppelin/contracts/token/ERC721/IERC721.sol",
  "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol",
  "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol",
  "@openzeppelin/contracts/utils/Address.sol",
  "@openzeppelin/contracts/utils/Context.sol",
  "@openzeppelin/contracts/utils/Counters.sol",
  "@openzeppelin/contracts/utils/Strings.sol",
  "@openzeppelin/contracts/utils/introspection/ERC165.sol",
  "@openzeppelin/contracts/utils/introspection/IERC165.sol",
  "Test.sol",
]
`;
